moodle/podcontroller.go

```diff
@@ -0,0 +1,242 @@
+	// 	return fmt.Errorf("failed to wait for caches to sync")
+	// }
+
+	glog.Info("Starting workers")
+
+	// Launch two workers to process Pod resources
+	for i := 0; i < threadiness; i++ {
+		go wait.Until(c.runWorker, time.Second*20, stopCh)
+	}
+	fmt.Println("c")
+	glog.Info("Started workers")
+	<-stopCh
+	glog.Info("Shutting down workers")
+
+	return nil
+}
+
+// runWorker is a long-running function that will continually call the
+// processPodQueueItem function in order to read and process a message on the
+// workqueue.
+func (c *PodController) runWorker() {
+	for c.processPodQueueItem() {
+	}
+}
+func (c *PodController) processPodQueueItem() bool {
+	podObj, shutdown := c.podqueue.Get()
+	if shutdown {
+		return false
+	}
+	err := func(podObj interface{}) error {
+		defer c.podqueue.Done(podObj)
+		var key string
+		var ok bool
+		if key, ok = podObj.(string); !ok {
+			c.podqueue.Forget(podObj)
+			runtime.HandleError(fmt.Errorf("PodController.go: expected string in podqueue but got %#v", podObj))
+			return nil
+		}
+		if err := c.handlePod(key); err != nil {
+			return fmt.Errorf("PodController.go: error in handlePod key: %s, err: %s", key, err.Error())
+		}
+		c.podqueue.Forget(podObj)
+		return nil
+	}(podObj)
+
+	if err != nil {
+		runtime.HandleError(err)
+		return true
+	}
+	return true
+}
+
+// enqueuePod takes a Pod resource and converts it into a namespace/name
+// string which is then put onto the work queue. This method should *not* be
+// passed resources of any type other than Foo.
+func (c *PodController) enqueuePod(obj interface{}) {
+	var key string
+	var err error
+	if key, err = cache.MetaNamespaceKeyFunc(obj); err != nil {
+		runtime.HandleError(err)
+		return
+	}
+	fmt.Printf("PodController.go     : enqueuePod, adding a pod key %s\n", key)
+	c.podqueue.AddRateLimited(key)
+}
+func (c *PodController) handlePod(key string) error {
+	// Convert the namespace/name string into a distinct namespace and name
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		runtime.HandleError(fmt.Errorf("PodController.go: invalid resource key, can't split : %s", key))
+		return fmt.Errorf("PodController.go: invalid resource key, can't split : %s", key)
+	}
+	pod, err := c.kubeclientset.CoreV1().Pods(namespace).Get(name, metav1.GetOptions{})
+	if err != nil {
+		// fmt.Println("PodController.go     : Couldn't find the replicaset, skipping")
+		return fmt.Errorf("Could not find the pod for pod name : %s, namespace %s", name, namespace)
+	}
+	ownerReferences := pod.GetOwnerReferences()
+	if len(ownerReferences) == 0 {
+		// fmt.Println("PodController.go     : Pod does not have any owner references.")
+		return nil
+	}
+	ownerReferenceName := ownerReferences[0].Name
+
+	replicaSet, err := c.kubeclientset.AppsV1().ReplicaSets(namespace).Get(ownerReferenceName, metav1.GetOptions{})
+	if err != nil {
+		// fmt.Println("PodController.go     : Couldn't find the replicaset, skipping")
+		return nil
```

Comment:
return the formatted error here -- similar to how you are doing it in line 173.

Reply:
The reason why I produce an error on line 173, is because when calling this method, the name/namespace better refer to a pod that exists, since that is what PodController is watching. So if this is called and I cannot find a Pod, then some error happened between the time it was pushed to the queue, popped off, parsed, and then called here.

As for this comment and the one below. This function is a filter -- filtering out all of the pods that do not have OwnerReference as Moodle. Because we do not want to act or even care about the pods that do not look like pod -> replicaset->deployment->Moodle. So this is intentional,  the idea is that if this deployment ownerreference does not exist, then hence we should just skip this Pod because it is not a Moodle. If I were to a return an error here, I would be printing a lot of errors to console (err:=handlePod(..); err!=nil{ print .... }) that really aren't errors in the first place

Reply:
Good explanation! Makes sense.